### PR TITLE
fix: Issue #1444 - API import with drop=true deletes existing client keys

### LINKED ISSUE #1444: API import with `drop=true` deletes existing client keys
Reported in the Slack community:

When importing data via the [API import functionality](https://docs.getunleash.io/api/admin/state#import-feature-toggles--strategies), specifying the `drop` query parameter will also drop existing client keys.

After speaking with @ivarconr , this seems to be because the `drop` option causes Unleash to do a cascading delete on existing environments, thus deleting existing environment-based client keys. 

This is undocumented and probably not the intended behavior. We should investigate this and find a solution for this.

### PR DESCRIPTION
**Pull request by szczepad**
## About the changes
The goal of this PR was to not drop existing api-tokens on an import with the drop=true flag.
The reason why these api-tokens have been dropped was the fact, that environments serve as foreign-key on the api-tokens table. This foreign key was setup in a way in which a deletion of an environment (which is part of drop=true) would cascade into api-tokens.

To solve this issue the constraint described above was removed. This leads to api-tokens never being deleted if drop=true is set.
In order to not store unecessary api-tokens a check during import of configuration was added, which removes all tokens which do not belong to an existing environment.

Closes #1444  .

### Important files
The main files of the PR are the database migration in 

The main files of this PR are 
- the database-migration in src/migrations/20220528143630-dont-cascade-environment-deletion-to-apitokens.js , which handle not deleting the api-keys on an import with drop=true
- the changes in src/lib/services/state-service.ts which handle the cleanup of api-tokens after an import.

### CODE REVIEW

File: src/test/e2e/api/admin/state.e2e.test.ts [hunk 1]
```diff
@@ -337,3 +338,105 @@ test(`should import segments and connect them to feature strategies`, async () =
...
+    });
+
+    await app.request
+        .post('/api/admin/state/import?drop=true')
+        .attach('file', 'src/test/examples/v3-minimal.json')
+        .expect(202);
+
+    const apiTokens = await app.services.apiTokenService.getAllTokens();
+
+    console.log(apiTokens);
```
Comment (thomasheartman) [inline src/test/e2e/api/admin/state.e2e.test.ts:439]: Is this console log intentional or left in by mistake? Unless it serves a specific purpose, we probably want to remove it 💁🏼

File: src/migrations/20220528143630-dont-cascade-environment-deletion-to-apitokens.js [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+          ALTER TABLE api_tokens DROP CONSTRAINT api_tokens_environment_fkey;
+      `,
+        cb,
+    );
+};
+
+exports.down = function (db, cb) {
+    db.runSql(
+        `
+          ALTER TABLE api_tokens DROP CONSTRAINT api_tokens_environment_fkey;
```
Comment (thomasheartman) [inline src/migrations/20220528143630-dont-cascade-environment-deletion-to-apitokens.js:15]: Should both these commands be the same? If the up command drops the constraint, shouldn't the down command create that constraint?

**Review by thomasheartman — COMMENTED** [review]
Thanks so much for checking out the viability of this and finding a way to make it work! 🙌🏼  I'm generally in favor of this change, but dropping the DB constraints gives me pause. I'll loop in someone else with more insight into the db structure, so that we can work through this together ☺️  I've added a few comments inline that I think may be small errors; feel free to tell me if they're not!

**Comment by chriswk** [comment]
Having looked through this, and seeing that you are indeed still deleting tokens for environments that did not get readded with the import. I tend towards a :+1: . As Thomas said earlier, thanks so much for taking the time to making the effort.

**Review by chriswk — CHANGES_REQUESTED** [review]
I'm missing a test for the environment service. Does api-tokens still get deleted if I delete an environment using the environment-service, rather than during the import/export code?

**Comment by szczepad** [comment]
Thanks for the valuable Feedback! :)
Indeed I'm currently not checking on the environment-service. Will add that quickly. :)

**Comment by szczepad** [comment]
Hey @chriswk,

> I'm missing a test for the environment service. Does api-tokens still get deleted if I delete an environment using the environment-service, rather than during the import/export code?

I just had a look on the environment-service. It seems like currently there's no functionality to delete an environment exposed by that service, but only to remove environments from projects. In the latter I suppose it wouldn't make sense to delete the api-tokens as the correlating environment should still be in the db and usable.

Should I add the functionality to delete environments, to avoid unexpected behaviour regarding the deletion of api-tokens in the future? If not, all other remarks for this PR should be resolved.

Thanks! :)

**Comment by chriswk** [comment]
Agreed, that removing an environment from a project is not enough to warrant removing the API key. I won't expect you to expand environment-service, so I'd say we're good to go.

**Review by chriswk — APPROVED** [review]
Thank you for this. :)

**Pull request merged**